### ISSUE TITLE
Given 2 ASNs, tell me where they overlap

### ISSUE THREAD

**Issue by mcmanuss8** [peeringdb/peeringdb#16]
Use cases:

1) As an operator, I see significant traffic to an ASN and I want to peer with them. I don't know what IXes we are both at. 

2) As an operator, I have public peering with an ASN, and I want to move to PNI because traffic volumes are high enough now. I want to see what facilities we have presence in besides the public ones.

3) As an operator, I don't have any common facilities with an ASN, but we're in the same metros. I want to see how difficult it would be to get from the facility/facilities I'm in in $metro to the ones they're in in $metro. This bleeds into https://github.com/peeringdb/peeringdb/issues/11. 

Use case 3 is harder, and we need more data about facility <-> facility interconnection or equivalence to do properly. But 1 and 2 should be easy. 

You could imagine showing this when logged in as an ASN. You browse to https://www.peeringdb.com/net/$X and you see where the net you're affiliated with overlaps with $X. 

Something like this seems to work (but doesn't handle local_asn so could use some improvement...): 

```
select distinct * from
(select 
    na.name isp_1, 
    nb.name isp_2, 
    fa.name, 
   fa.city city,
    fa.country country,
    'private' connection_type
from 
    net na, 
    net nb, 
    netfac fa, 
    netfac fb
where 
    na.id = fa.net_id 
    and nb.id = fb.net_id 
    and na.asn = $asn_1 
    and nb.asn = $asn_2 
    and fa.fac_id = fb.fac_id
union all
select 
    na.name isp_1, 
    nb.name isp_2, 
    ix.name, 
    ix.city city,
    ix.country country,
    'public' connection_type
from 
    net na, 
    net nb, 
    netixlan ila, 
    netixlan ilb,
    ix
where 
    na.id = ila.net_id 
    and nb.id = ilb.net_id 
    and na.asn = $asn_1 
    and nb.asn = $asn_2 
    and ila.ix_id = ilb.ix_id 
    and ix.id = ila.ix_id
) z
```

**Comment by vegu** [comment]
> You could imagine showing this when logged in as an ASN. You browse to https://www.peeringdb.com/net/$X and you see where the net you're affiliated with overlaps with $X.

Users may be associated with multiple networks, and at some point this is just going to become messy. Also probably not a query we want to do every-time someone loads a network view.

Instead i suggest we implement this as a functionality of the API, build a form for it in the advanced search tool, and maybe add a link to that to the network page for convenience. Also this is assuming that anyone should be able to an overlap query between two ASNs, whether or not they actually are a member of either.

**Comment by eloos** [comment]
agreed with the solution/analysis of @vegu

**Comment by mcmanuss8** [comment]
I agree you don't necessarily want to load this on every load if that's taxing on the db. I would like to see the concept of optional widgets though. So if there's a data set that's interesting but not used 100% of the time, you can click it to expand and it'll load ajaxly. 
New ticket for that that blocks this one, @vegu  ?

**Comment by vegu** [comment]
That should definitely be a separate ticket, yes. I am not sure it needs to block this one especially if this one is treated as a query feature of the API with a simple form in advanced search for the time being. Or even just have it as an api feature, if having it in advanced search seems overboard.

We can always make ui widgets around the api at a later point, and focus on the functionality aspect for now.

**Comment by mcmanuss8** [comment]
Ok opened #265 for what we could do to make the ui better for this case. Let's make it an API feature then expose it on /net/ as an optional widget once that's done.

**Comment by vegu** [comment]
Implemented API query for this in `89e7f4e7`

`/api/ix?asn_overlap=XXX,XXX,XXX,...`
`/api/fac?asn_overlap=XXX,XXX,XXX,...`

@mcmanuss8 @eloos - just so we are all on the same page, am i understanding correctly that a form element for this in the advanced search tool is currently not needed, and it's sufficient that an ui element will be provided via #265 at a later time.

**Comment by mcmanuss8** [comment]
@vegu  awesome.  agreed on the adding it as a ui element post #265
